File: examples/golang/9_1_phil.go
/**
 * (c) Ivan Janevski
 */
package main

import (
    "fmt"
    "time"
)

// constants
const N = 5

// channels
var forks [N]chan bool

// routines
func phil(id int, done chan bool) {
    left := id
    right := (id + 1) % N
    for {
        // Think
        fmt.Println("Philosopher", id, "is thinking")
        time.Sleep(time.Microsecond)
        // Acquire forks
        if id % 2 == 0 {
            <- forks[left]
            <- forks[right]
        } else {
            <- forks[right]
            <- forks[left]
        }
        // Eat
        fmt.Println("Philosopher", id, "is eating")
        time.Sleep(time.Microsecond)
        // Release forks
        forks[left] <- true
        forks[right] <- true
    }
    done <- true
}

func main() {
    // init channels
    done := make(chan bool, 1)
    for i := 0; i < N; i++ {
        forks[i] = make(chan bool, 1)
        forks[i] <- true
    }

    // run
    for i := 0; i < N; i++ {
        go phil(i, done)
    }

    // join
    for i := 0; i < N; i++ {
        <- done
    }
}

File: examples/clinda/9_1_phil.cpp
/**
 * (c) Ivan Janevski
 */
#include <print.h>
#include <clinda.h>
#define N 5


void phil(int id) {
    int left = id;
    int right = (id + 1) % N;
    while (true) {
        // Think
        println("Thinking %d", id);
        usleep(10);
        // Acquire forks
        if (id % 2 == 0) {
            in("fork", left);
            in("fork", right);
        } else {
            in("fork", right);
            in("fork", left);
        }
        // Eating
        println("Eating %d", id);
        usleep(10);
        // Release forks
        out("fork", left);
        out("fork", right);
    }
}

void init() {
    for (int i = 0; i < N; i++) 
        out("fork", i);
    for (int i = 0; i < N; i++) 
        eval(phil(i));
}
